### PR TITLE
fix(DepartureList): (subway only) remove schedules when showing predictions

### PR DESCRIPTION
**Pull request by thecristen**
#### Summary of changes

**Asana Ticket:** [Departures | Inaccurate subway times](https://app.asana.com/0/385363666817452/1205064463145400/f)

~~Adds another flag to the schedules endpoint, to adjust the returned schedules to exclude subway schedules unless they're from a terminus on the Green Line. Deploying to dev-blue for testing.~~

**The change in this PR applies to the `<DepartureList />` component only.** Another PR applying this logic to the landing page `<DepartureTimes />` and its rows of times... will come after.

This covers a revised approach where, **for subways only**, we suppress schedule-only departures in favor of predictions. When there are no remaining predictions to display (usually after an hour or two away), the list allows showing schedules again through the remainder of the day.

---

#### General checks
* [x] **Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?** This is a good practice that can facilitate easier reviews.
* [x] **Testing.** Do the changes include relevant passing updates to tests? This includes updating screenshots. Preferably tests are run locally to verify that there are no test failures created by these changes, before opening a PR.
* [ ] **Tech debt.** Have you checked for tech debt you can address in the area you're working in? This can be a good time to address small issues, or create Asana tickets for larger issues.

### CODE REVIEW

**Comment by kotva006** [comment]
This makes sense for what it is doing.  I thought we were using predictions over schedules on the display logic already.

File: apps/site/assets/ts/helpers/departureInfo.tsx [hunk 1]
```diff
@@ -123,11 +135,54 @@ const isCommuterRail = (departureInfo: DepartureInfo): boolean => {
...
+ * This function can optionally
+ * - wrap each `<DisplayTime />` in a specified wrapper element, `WrapperEl`
+ * - truncate the list to a desired `listLength` value
+ */
+const departuresListFromInfos = (
+  departureInfos: DepartureInfo[],
+  isCR: boolean,
+  targetDate?: Date,
+  listLength?: number,
+  WrapperEl: typeof DefaultWrapper = DefaultWrapper
```
Comment (kotva006) [inline apps/site/assets/ts/helpers/departureInfo.tsx:154]: Why is this needed?

Comment (thecristen) [reply apps/site/assets/ts/helpers/departureInfo.tsx:154]: @kotva006 Does the comment on the line above this function declaration help answer this at all? What details are you looking for?

Comment (kotva006) [reply apps/site/assets/ts/helpers/departureInfo.tsx:154]: I understand what it does, I just don't understand the point.  As this functionality doesn't appear to be used on this pull request

Comment (thecristen) [reply apps/site/assets/ts/helpers/departureInfo.tsx:154]: @kotva006 looks like we commented at the same time - see my second comment

File: apps/site/assets/ts/helpers/departureInfo.tsx [hunk 2]
```diff
@@ -123,11 +135,54 @@ const isCommuterRail = (departureInfo: DepartureInfo): boolean => {
...
+): React.ReactElement[] => {
+  // optional cutoff time, before which we won't show schedules.
+  // just used with subway for now.
+  const predictionTimeCutoff = chain(departureInfos)
+    .filter(d => d.routeMode === SUBWAY)
+    .maxBy("prediction.time")
+    .value()?.prediction?.time;
```
Comment (kotva006) [inline apps/site/assets/ts/helpers/departureInfo.tsx:161]: nit-pick(non-blocking): Since we are maxing by `"prediction.time"` I would think the `prediction` and `time` field would be defined, so they could probably be `prediction!.time`;  Though not sure if an `!` is allowed after a `?`

File: apps/site/assets/ts/helpers/departureInfo.tsx [hunk 3]
```diff
@@ -123,11 +135,54 @@ const isCommuterRail = (departureInfo: DepartureInfo): boolean => {
...
+        schedule &&
+        schedule.time <= predictionTimeCutoff
+    )
+    .map(d => (
+      <WrapperEl key={getInfoKey(d)}>
+        <DisplayTime departure={d} isCR={isCR} targetDate={targetDate} />
+      </WrapperEl>
+    ))
+    .slice(0, listLength)
+    .value();
```
Comment (kotva006) [inline apps/site/assets/ts/helpers/departureInfo.tsx:177]: Why is this complex `chain` logic needed over just using the `prediction` from the `departureInfos`?  Based of the ticket it seems like the goal is just to show predictions if they exist, which I thought we were doing already.
The `departureInfoToTime` function returns the prediction time in favor of the schedule time.
Maybe I am having a hard time following this logic.

Comment (thecristen) [reply apps/site/assets/ts/helpers/departureInfo.tsx:177]: @kotva006 Here's a non-chained translation, if it's helpful 🤔 

```javascript
const departuresListWithoutIntermingledSchedules = omitBy(departureInfos, departure => {
  if (predictionTimeCutoff) {
    // omit schedule-only departures that are before the latest prediction time
    !prediction && schedule && schedule.time <= predictionTimeCutoff;
  }
  return false;
});
const departureElements = departuresListWithoutIntermingledSchedules.map(d => /* render as <DisplayTime /> and wrap in wrapper element */)
return departureElements.slice(0, listLength);
```

File: apps/site/assets/ts/helpers/departureInfo.tsx (hunk 2 above)
Comment (thecristen) [reply apps/site/assets/ts/helpers/departureInfo.tsx:161]: I think you're right, and I just tried it and it works.

**Review by kotva006 — APPROVED** [review]
Looks fine
